iceoryx_hoofs/include/iceoryx_hoofs/design_pattern/static_lifetime_guard.hpp

```diff
@@ -0,0 +1,125 @@
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_DESIGN_STATIC_LIFETIME_GUARD_HPP
+#define IOX_HOOFS_DESIGN_STATIC_LIFETIME_GUARD_HPP
+
+#include <atomic>
+#include <type_traits>
+#include <utility>
+
+namespace iox
+{
+namespace design_pattern
+{
+/// @brief Manages a static instance of type T in a way so that each
+/// existing StaticLifetimeGuard prevents the destruction of
+/// the instance.
+/// 1. instance() creates a static guard itself and hence has static lifetime
+/// 2. any static StaticLifetimeGuard G created before that prolongs the lifetime
+/// of the instance at least until G is destroyed
+/// @tparam T the type of the instance to be guarded
+///
+/// @note all public functions except setCount are thread-safe
+/// @attention setCount is only supposed to be used while no guards are created
+/// concurrently
+///
+/// @code
+/// // instance will be destroyed after guard
+/// // (or later if there are guards preceding guard in construction order)
+///
+/// static StaticLifetimeGuard<T> guard;
+/// static T& instance = StaticLifetimeGuard<T>::instance();
+/// @endcode
+template <typename T>
+class StaticLifetimeGuard
+{
+  public:
+    StaticLifetimeGuard() noexcept;
+
+    StaticLifetimeGuard(const StaticLifetimeGuard&) noexcept;
+
+    StaticLifetimeGuard(StaticLifetimeGuard&&) noexcept;
+
+    ~StaticLifetimeGuard() noexcept;
+
+    // assignment does nothing since the objects have no state,
+    // copy and move exist to support passing/returning a value object
+    StaticLifetimeGuard& operator=(const StaticLifetimeGuard&) noexcept = default;
+    StaticLifetimeGuard& operator=(StaticLifetimeGuard&&) noexcept = default;
+
+    /// @brief Construct the instance to be guarded with constructor arguments.
+    /// @param args constructor arguments
+    /// @return reference to the constructed instance or the existing instance
+    /// if it already exists
+    /// @note creates an implicit static StaticLifetimeGuard to ensure
+    /// the instance is destroyed in the static destruction phase
+    template <typename... Args>
+    static T& instance(Args&&... args) noexcept;
+
+    /// @brief Set the instance life time count.
+    /// @param count value to be set
+    /// @return previous count value
+    /// @note This can be used to additionally extend or shorten the instance lifetime,
+    /// This has to be done carefully, to ensure destruction or prevent early destruction.
+    /// It is useful for testing purposes.
+    /// @attention setCount is only supposed to be used while no guards are created
+    /// concurrently as it will influence the guard mechanism.
+    /// The instance is destroyed once the counter reaches zero upon destruction of a guard,
+    /// i.e. changing the counter may lead to the instance never being destroyed
+    /// or being destroyed before the last guard.
+    static uint64_t setCount(uint64_t count);
```

Comment:
Can be done but I dislike the naming of the method and would prefer marking only. Also because there is no clear and agreed definition what unsafe means (I could provide one, but would mean something different). The problem is that the function itself is not unsafe as far as it is used correctly. It only exists for testing, I would prefer it did not even exist.

Using the type system is possible but quit a bit of overhead I would say. I think about it.

Reply:
If it only exists for testing, it could be made private and a friend to e.g. `StaticLifetimeGuardTestAccessor` could be made to access the method in tests. We already did similar things for other classes

Reply:
Made protected and only used for testing by inheritance (athis will not alter the object here in any case, as we only access a static method). This is less annoying than dealing with friends!